AmplifyTools/AmplifyXcode/Sources/AmplifyXcodeCore/Commands/AmplifyCommand.swift

```diff
@@ -0,0 +1,26 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// An `AmplifyCommand` describes the shape of a command. A command is just an abstraction and doesn't
+/// make any assumptions about neither the environment in which will be executed or the executor interface.
+/// It's the executor's responsibility to provide the necessary arguments to the initializer of a conforming type.
+///
+/// - `taskArgs`: type defining extra arguments that will be passed to each task
+/// - `tasks`: array of tasks whose a command is composed of
+/// - `description`: human-readable description of command purpose
+public protocol AmplifyCommand {
+    associatedtype TaskArgs
+    var taskArgs: TaskArgs { get }
+
+    var tasks: [AmplifyCommandTask<TaskArgs>] { get }
+
+    static var description: String { get }
+
+    func onFailure()
```

Comment:
Not sure any action is needed now, but I'll note it's surprising to me to see an `onFailure` callback with no corresponding `onSuccess`.

Reply:
I considered to add an `onSuccess` callback but I couldn't find an immediate usage, but I suspect will be handy for other commands.